python/cuda_parallel/tests/test_reduce.py

```diff
@@ -550,3 +550,107 @@ def binary_op(x, y):
     d_in = cp.zeros(size)[::2]
     with pytest.raises(ValueError, match="Non-contiguous arrays are not supported."):
         _ = algorithms.reduce_into(d_in, d_out, binary_op, h_init)
+
+
+def test_reduce_with_stream():
+    # Simple cupy stream wrapper that implements the __cuda_stream__ protocol for the purposes of this test
+    class Stream:
+        def __init__(self, cp_stream):
+            self.cp_stream = cp_stream
+
+        def __cuda_stream__(self):
+            return (0, self.cp_stream.ptr)
+
+    def add_op(x, y):
+        return x + y
+
+    h_init = np.asarray([0], dtype=np.int32)
+    h_in = random_int(5, np.int32)
+
+    stream = cp.cuda.Stream()
+    with stream:
+        d_in = cp.asarray(h_in)
+        d_out = cp.empty(1, dtype=np.int32)
+
+    stream_wrapper = Stream(stream)
+    reduce_into = algorithms.reduce_into(
+        d_in=d_in, d_out=d_out, op=add_op, h_init=h_init
+    )
+    temp_storage_size = reduce_into(
+        None,
+        d_in=d_in,
+        d_out=d_out,
+        num_items=d_in.size,
+        h_init=h_init,
+        stream=stream_wrapper,
+    )
+    d_temp_storage = cp.empty(temp_storage_size, dtype=np.uint8)
```

Comment:
This is technically a bug that could surface in a large test suite: CuPy is backed by a stream-ordered allocator so here we allocate memory on the default stream and use it on the custom stream, without proper ordering. It could be possible that this memory is still being accessed by a previous kernel on the default stream (despite it's been returned to the allocator!), so if the next (CUB) kernel is launched in parallel on the custom stream there's a race condition.

The safest option is to allocate memory in CuPy's stream context.

Reply:
(Note that I said "default stream" without mentioning if it's the legacy one or the PTDS, since it's a behavior we should not rely upon.)